### PR TITLE
Check public key duplication - Closes #489

### LINKED ISSUE #489: Add check for passed public key multiplication in votes and keysgroup
To get sure the public keys are not only sane (https://github.com/LiskHQ/lisk-js/issues/488) but also do not contain duplicates.

### PR DESCRIPTION
**Pull request by Tosch110**
Closes #489
~~Merge https://github.com/LiskHQ/lisk-js/pull/504 first~~

### CODE REVIEW

File: test/transactions/3_castVotes.js [hunk 1]
```diff
@@ -164,4 +164,104 @@ describe('#castVotes transaction', () => {
...
+			});
+		});
+
+		it('the transaction should have the votes object', () => {
+			castVotesTransaction.asset.should.have
+				.property('votes')
+				.and.be.type('object');
+		});
+
+		it('the transaction should have two upvotes and two unvotes', () => {
```
Comment (willclarktech) [inline test/transactions/3_castVotes.js]: Incorrect description

File: test/transactions/3_castVotes.js [hunk 2]
```diff
@@ -164,4 +164,104 @@ describe('#castVotes transaction', () => {
...
+			castVotes
+				.bind(null, {
+					passphrase,
+					votes,
+				})
+				.should.throw(
+					'Duplicated public key: 5d036a858ce89f844491762eb89e2bfbd50a4a0a0da658e4b2628b25b117ae09.',
+				);
+		});
+	});
```
Comment (willclarktech) [inline test/transactions/3_castVotes.js]: What about a duplicate in unvotes?

File: test/transactions/3_castVotes.js [hunk 3]
```diff
@@ -164,4 +164,104 @@ describe('#castVotes transaction', () => {
...
+					passphrase,
+					unvotes,
+					votes,
+				})
+				.should.throw(
+					'Duplicated public key: 5d036a858ce89f844491762eb89e2bfbd50a4a0a0da658e4b2628b25b117ae09.',
+				);
+		});
+
+		it('should throw an error', () => {
```
Comment (willclarktech) [inline test/transactions/3_castVotes.js]: These tests need distinct titles.

File: test/transactions/3_castVotes.js [hunk 4]
```diff
@@ -164,4 +164,104 @@ describe('#castVotes transaction', () => {
...
+	describe('when the cast vote transaction is created with the votes and unvotes', () => {
+		beforeEach(() => {
+			castVotesTransaction = castVotes({
+				passphrase,
+				votes: votePublicKeys,
+				unvotes: unvotePublicKeys,
+			});
+		});
+
+		it('the transaction should have the votes object', () => {
```
Comment (willclarktech) [inline test/transactions/3_castVotes.js]: I'm not sure we need this if we have the test below it, but if we do isn't it an array? There's a should.js assertion for that.

File: test/transactions/3_castVotes.js [hunk 5]
```diff
@@ -164,4 +164,104 @@ describe('#castVotes transaction', () => {
...
+
+	describe('when the cast vote transaction is created with the unvotes', () => {
+		beforeEach(() => {
+			castVotesTransaction = castVotes({
+				passphrase,
+				unvotes: unvotePublicKeys,
+			});
+		});
+
+		it('the transaction should have the votes object', () => {
```
Comment (willclarktech) [inline test/transactions/3_castVotes.js]: As above.

File: src/transactions/3_castVotes.js [hunk 6]
```diff
@@ -18,22 +18,44 @@
...
 
-const castVotes = ({ passphrase, delegates, secondPassphrase, timeOffset }) => {
+const castVotes = ({
+	passphrase,
+	votes,
+	unvotes,
+	secondPassphrase,
+	timeOffset,
+}) => {
+	hasNoDuplication([...(votes || []), ...(unvotes || [])]);
```
[the remark this answers is not in the document]
Comment (Tosch110) [reply src/transactions/3_castVotes.js]: The reason why it does not have a default value is that both functions (validatePublicKey and prependPlusToPublicKeys would need to have an extra boilerplate in order to work with an empty array.

File: src/transactions/utils/validation.js [hunk 7]
```diff
@@ -28,8 +28,18 @@ export const validatePublicKey = publicKey => {
 	return true;
 };
 
+export const checkPublicKeysForDuplicates = publicKeys =>
+	publicKeys.every((element, index) => {
+		const elementFound = publicKeys.indexOf(element);
```
Comment (willclarktech) [inline src/transactions/utils/validation.js]: How about `publicKeys.slice(index + 1)`? Makes it slightly more efficient and enables a simpler `if` check below.

**Pull request merged**